File: src/suite_actuarial/api/telemetry.py
"""Privacy-safe, opt-in server-side telemetry for API request health."""

from __future__ import annotations

import asyncio
import json
import logging
import os
import time
from collections.abc import Mapping
from urllib.request import Request, urlopen

LOGGER = logging.getLogger(__name__)
POSTHOG_EVENT = "api_request"
POSTHOG_DISTINCT_ID = "suite-actuarial-api"


def _posthog_api_key() -> str:
    return os.environ.get("POSTHOG_PROJECT_API_KEY", os.environ.get("POSTHOG_API_KEY", "")).strip()


def posthog_enabled() -> bool:
    """Return whether the deployment has opted into PostHog capture."""
    return bool(_posthog_api_key())


def build_api_event_payload(
    *,
    route: str,
    method: str,
    status_code: int,
    duration_ms: int,
    outcome: str,
    api_key: str,
) -> dict[str, object]:
    """Build the narrow event contract sent to PostHog.

    Request bodies, query strings, response bodies, user agents, IP addresses,
    and actuarial inputs/results are intentionally excluded.
    """
    properties: dict[str, object] = {
        "distinct_id": POSTHOG_DISTINCT_ID,
        "$process_person_profile": False,
        "$lib": "suite_actuarial_api",
        "$lib_version": "2.2.0",
        "api_route": route,
        "http_method": method,
        "status_code": int(status_code),
        "duration_ms": max(0, int(duration_ms)),
        "outcome": outcome,
    }
    return {"api_key": api_key, "event": POSTHOG_EVENT, "properties": properties}


def _send_payload(payload: Mapping[str, object]) -> None:
    host = os.environ.get("POSTHOG_HOST", "https://us.i.posthog.com").rstrip("/")
    request = Request(
        f"{host}/capture/",
        data=json.dumps(payload, separators=(",", ":")).encode("utf-8"),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    with urlopen(request, timeout=1.5) as response:  # noqa: S310 - host is deployment configuration
        response.read()


def _log_task_failure(task: asyncio.Task[object]) -> None:
    try:
        task.result()
    except Exception:  # pragma: no cover - network failures depend on deployment
        LOGGER.debug("PostHog capture failed", exc_info=True)


def schedule_api_event(
    *,
    route: str,
    method: str,
    status_code: int,
    started_at: float,
) -> None:
    """Queue a capture without delaying or failing the API response."""
    api_key = _posthog_api_key()
    if not api_key:
        return
    duration_ms = round((time.perf_counter() - started_at) * 1000)
    outcome = (
        "success" if status_code < 400 else "client_error" if status_code < 500 else "server_error"
    )
    payload = build_api_event_payload(
        route=route,
        method=method,
        status_code=status_code,
        duration_ms=duration_ms,
        outcome=outcome,
        api_key=api_key,
    )
    try:
        task = asyncio.create_task(asyncio.to_thread(_send_payload, payload))
    except RuntimeError:  # pragma: no cover - only possible without an active event loop
        return
    task.add_done_callback(_log_task_failure)

File: edge/src/telemetry.ts
/**
 * PostHog capture from the edge.
 *
 * This mirrors the contract in `src/suite_actuarial/api/telemetry.py` and
 * deliberately keeps the same shape, but it is a separate event
 * (`api_edge_request`) rather than a second source for `api_request`. Two
 * reasons: the origin never sees a request that was rate limited or served
 * from cache, so the counts would not agree; and keeping them distinct means a
 * discrepancy between the two is readable as "the edge absorbed it" instead of
 * looking like data loss.
 *
 * What is not sent, matching the project's standing analytics contract: request
 * bodies, query strings, response bodies, actuarial inputs or results, user
 * agents, IP addresses, and API keys. Country and Cloudflare location are new
 * to the edge event and are documented in `docs/ANALYTICS.md`.
 */

import type { CacheState, Caller, Outcome } from "./types";

export const EDGE_EVENT = "api_edge_request";
export const ANONYMOUS_DISTINCT_ID = "suite-actuarial-api-edge";
const LIB_VERSION = "2.2.0";

export function outcomeFor(status: number): Outcome {
  if (status === 429) return "rate_limited";
  if (status < 400) return "success";
  if (status < 500) return "client_error";
  return "server_error";
}

export interface EdgeEventInput {
  route: string;
  method: string;
  status: number;
  durationMs: number;
  cache: CacheState;
  caller: Caller;
  country?: string;
  colo?: string;
  apiKey: string;
}

export function buildEdgeEventPayload(input: EdgeEventInput): Record<string, unknown> {
  const properties: Record<string, unknown> = {
    // A fixed id for anonymous traffic, the key label for identified traffic.
    // Person profiles stay off in both cases: the point is to see which
    // surfaces are used and by which integration, not to build a profile.
    distinct_id: input.caller.tier === "key" ? `key:${input.caller.label}` : ANONYMOUS_DISTINCT_ID,
    $process_person_profile: false,
    $lib: "suite_actuarial_edge",
    $lib_version: LIB_VERSION,
    api_route: input.route,
    http_method: input.method,
    status_code: input.status,
    duration_ms: Math.max(0, Math.round(input.durationMs)),
    outcome: outcomeFor(input.status),
    cache: input.cache,
    auth_tier: input.caller.tier,
  };
  if (input.caller.tier === "key") properties.api_key_label = input.caller.label;
  if (input.country) properties.country = input.country;
  if (input.colo) properties.colo = input.colo;

  return { api_key: input.apiKey, event: EDGE_EVENT, properties };
}

/**
 * Send the event. Never throws: a telemetry outage must not become an API
 * outage, and the caller has already received their response by the time this
 * runs inside `waitUntil`.
 */
export async function captureEdgeEvent(
  host: string,
  payload: Record<string, unknown>,
): Promise<void> {
  try {
    await fetch(`${host.replace(/\/+$/, "")}/capture/`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
  } catch (error) {
    console.error("PostHog capture failed", error);
  }
}
